### PR TITLE
Updated Dockefile for crosscompilation compatibility

### PR DESCRIPTION
**Pull request by OlivierCazade**
This update the Dockerfile to make it possible to cross build FLP container image.

Example of manual cross build command:
`podman buildx build --platform=linux/ppc64le -t test -f ./contrib/docker/Dockerfile .`

Could someone with docker installed could confirm that the cli API is identical and the example above also works with docker?

### CODE REVIEW

**Review by mariomac — APPROVED** [review]
LGTM

**Review by mariomac — CHANGES_REQUESTED** [review]
Unapproving as I get this error in Rancher Desktop (docker) for Mac:

```
$ docker buildx build --platform=linux/ppc64le -t test -f ./contrib/docker/Dockerfile .
[+] Building 0.1s (2/2) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                     0.0s
 => => transferring dockerfile: 789B                                                                                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                                                                                        0.0s
 => => transferring context: 34B                                                                                                                                                                                         0.0s
ERROR: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to parse platform : "" is an invalid component of "": platform specifier component must match "^[A-Za-z0-9_-]+$": invalid argument
```

**Review by mariomac — COMMENTED** [review]
It seems the issue is related to Mac, so anyone else should try it as I can't test it

**Comment by jotak** [comment]
@OlivierCazade it works for me but I'm not using docker either
cc @ronensc @eranra @KalmanMeth does it work for you too, in your environment?

Also Olivier, we should ensure it works with the github CI .. can you try adding this build in some of the workflows ? E.g. just temporarily in `pull_request.yaml` (just for testing, after that we can remove it); And after that we should consider adding it to `release.yaml`

**Comment by jotak** [comment]
By the way we should still try to fix it for mac if possible (or create an issue to fix it later)

**Comment by KalmanMeth** [comment]
I tried to run with docker and got an error.

docker buildx build --platform=linux/ppc64le -t test -f ./contrib/docker/Dockerfile .
[+] Building 160.8s (8/21)
 => [internal] load build definition from Dockerfile                                                                     0.1s
 => => transferring dockerfile: 640B                                                                                     0.0s
 => [internal] load .dockerignore                                                                                        0.2s
 => => transferring context: 45B                                                                                         0.0s
 => [internal] load metadata for docker.io/library/golang:1.18                                                           5.0s
 => [internal] load metadata for registry.access.redhat.com/ubi8/ubi-minimal:8.7                                         2.8s
 => [internal] load build context                                                                                        2.5s
 => => transferring context: 323.14MB                                                                                    2.4s
 => CANCELED [builder  1/12] FROM docker.io/library/golang:1.18@sha256:00d63686b480f6dc866e93ddc4b29efa2db03274a687e6  155.6s
 => => resolve docker.io/library/golang:1.18@sha256:00d63686b480f6dc866e93ddc4b29efa2db03274a687e6495c2cfbfe615d638e     0.0s
 => => sha256:5363f0c14d79c77678586642d6ff9ef822416864a8358da267bef743bb215907 11.63MB / 11.63MB                         6.5s
 => => sha256:00d63686b480f6dc866e93ddc4b29efa2db03274a687e6495c2cfbfe615d638e 2.36kB / 2.36kB                           0.0s
 => => sha256:bdd0f8661015f601ef21ee0b43d6ac02e1b78b32bd5ceb1217efbe9f7f95d117 1.80kB / 1.80kB                           0.0s
 => => sha256:54d1c7d8aa9bd6a4eccbc8d2a6bfbff823a47ff684d332c8a50a2bb81dd94f1f 7.11kB / 7.11kB                           0.0s
 => => sha256:542880f4d15404af17d342ecbb76bb92724fc7cb8a91a5e18f26bd8f8811a38a 58.90MB / 58.90MB                        80.8s
 => => sha256:a9b0d1697a3bff7991f1c96b22ffd254326c700694ddcf45ba42cd27f8a9cb94 5.41MB / 5.41MB                           5.1s
 => => sha256:1ec399ef8d37e347a1f93e54dd49428d1cec8a8926b5b48c5823c1e9813d561a 58.87MB / 58.87MB                        96.5s
 => => sha256:641c1c2e99a77066b47db2913c8bb3d7c19ed2a95ea0cc897aac1501e3744ace 76.55MB / 80.44MB                       155.4s
 => => sha256:84324359b1058d26717ef39b7917b48fcb28409dcd5970d0e5214139494fd363 45.09MB / 109.13MB                      155.4s
 => => extracting sha256:542880f4d15404af17d342ecbb76bb92724fc7cb8a91a5e18f26bd8f8811a38a                                1.5s
 => => extracting sha256:a9b0d1697a3bff7991f1c96b22ffd254326c700694ddcf45ba42cd27f8a9cb94                                0.1s
 => => extracting sha256:5363f0c14d79c77678586642d6ff9ef822416864a8358da267bef743bb215907                                0.2s
 => => extracting sha256:1ec399ef8d37e347a1f93e54dd49428d1cec8a8926b5b48c5823c1e9813d561a                                1.6s
 => => sha256:3ffce554075f0c84744be2d3fec1c315f4684af55fccf8e238f1da1728610ae6 155B / 155B                              97.7s
 => [stage-1 1/4] FROM registry.access.redhat.com/ubi8/ubi-minimal:8.7@sha256:5d2d4d4dbec470f8ffb679915e2a8ae25ad754c  153.9s
 => => resolve registry.access.redhat.com/ubi8/ubi-minimal:8.7@sha256:5d2d4d4dbec470f8ffb679915e2a8ae25ad754cd9193fa966  0.1s
 => => sha256:5d2d4d4dbec470f8ffb679915e2a8ae25ad754cd9193fa966deee1ecb7b3ee00 1.47kB / 1.47kB                           0.0s
 => => sha256:520a368f78807947b96ea773cc62b14e380f4af08bbfd8ed18f0ebc70dedef68 429B / 429B                               0.0s
 => => sha256:53c1fc317e043407339a90c76d41cfb385f0bb1fe45e9f7180d916bfaf9f1e4f 5.98kB / 5.98kB                           0.0s
 => => sha256:c35a4f7d28053265fc2bc203599b2038ece031bb651bc7c0a3dfb05222daff37 43.94MB / 43.94MB                       152.7s
 => => extracting sha256:c35a4f7d28053265fc2bc203599b2038ece031bb651bc7c0a3dfb05222daff37                                0.8s
 => ERROR [stage-1 2/4] RUN microdnf install -y   iputils   curl   net-tools   && microdnf -y clean all  && rm -rf /var  1.6s
------
 > [stage-1 2/4] RUN microdnf install -y   iputils   curl   net-tools   && microdnf -y clean all  && rm -rf /var/cache:
#0 0.422 exec /bin/sh: exec format error
------
error: failed to solve: executor failed running [/bin/sh -c microdnf install -y   iputils   curl   net-tools   && microdnf -y clean all  && rm -rf /var/cache]: exit code: 1

**Comment by KalmanMeth** [comment]
My mistake. I used the wrong branch.
Using the branch from this PR I now get the following:

```
docker buildx build --platform=linux/ppc64le -t test -f ./contrib/docker/Dockerfile .
[+] Building 0.2s (2/2) FINISHED
 => [internal] load build definition from Dockerfile                                                                     0.1s
 => => transferring dockerfile: 789B                                                                                     0.0s
 => [internal] load .dockerignore                                                                                        0.1s
 => => transferring context: 45B                                                                                         0.0s
error: failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to parse platform : "" is an invalid component of "": platform specifier component must match "^[A-Za-z0-9_-]+$": invalid argument
```

**Comment by KalmanMeth** [comment]
Explicitly defining BUILDPLATFORM, etc gets past this error:
`docker buildx build --platform=linux/ppc64le --build-arg BUILDPLATFORM=linux/ppc64le --build-arg TARGETPLATFORM=linux/ppc64le -t test -f ./contrib/docker/Dockerfile `
but it then gets stuck on git.

**Comment by OlivierCazade** [comment]
I installed docker locally to check at differences, BUILDPLATFORM and TARGETPLATFORM are supposed to be defined automatically when setting the --platform option.

But it looks like there are some differences, podman expect you to define them with ARG but not docker.
It looks like both are happy if the ARG part is defined after the first FROM.

**Comment by OlivierCazade** [comment]
I added new Makefile target:
`DOCKER_IMG="quay.io/<user>/flowlogs-pipeline" DOCKER_TAG="test" make push-image-multiarch`

It should automatically build for arm64, amd64 and ppc64le, build the corresponding manifest and push it.

I tested on my computer with both podman and docker

**Comment by OlivierCazade** [comment]
Ok, this time it is even working on GH action, it is a bit sad that we can not use easy feature from podman and docker to cross build and generate the manifest because both are incompatible and it would force us to maintain two Dockerfile

**Comment by jotak** [comment]
@OlivierCazade cool!
So wdyt, better to keep the multi-arch build per-PR, or only during release?
It increases the build time quite a bit, and will grow when we add new archs; if we're confident enough that there little chance to see compatibility issues in FLP code, I'd say we can keep building only amd64 on github PRs
And we'll do the multi-arch in the release scripts
?

File: contrib/docker/Dockerfile [hunk 1]
```diff
@@ -1,5 +1,10 @@
-FROM docker.io/library/golang:1.18 as builder
+# TARGETPLATFORM and BUILDPLATFORM are automatically field when specifying an architecture during build but they need to be declared
```
Comment (KalmanMeth) [inline contrib/docker/Dockerfile:1]: field -> filled

Comment (OlivierCazade) [reply contrib/docker/Dockerfile]: Thanks, I changed a bit this comment to be more accurate with the last commits.

**Comment by OlivierCazade** [comment]
@jotak based on what we tested during the hackathon and assuming we do a first round of testing with new arch, I would agree to only build amd64 during PR and build other architecture during release.

**Review by mariomac — APPROVED** [review]
Tested locally with docker (rancher desktop) in a Mac

**Pull request merged**